### ISSUE TITLE
Add a Redis-backed token bucket RateLimiter implementation

### ISSUE THREAD

**Issue by halter73** [dotnet/aspnetcore#41861]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

In the runtime repo, we have included a bunch of built-in, in-memory `RateLimiter` implementations like `ConcurrencyLimiter`, `FixedWindowRateLimiter`, `SlidingWindowRateLimiter` and `TokenBucketRateLimiter`. It would be nice to add a rate limiter that works across multiple hosts in a distributed environment. In the past, for things like our `RedisCache` `IDistributedCache` implementation, we've used StackExchange.Redis to implement a solution in the aspnetcore repo for a runtime abstraction. I think we should do the same things for this.

### Describe the solution you'd like

I think we should implement a token bucket `RateLimiter` based on Redis using StackExchange.Redis that can be used in distributed environments. I imagine the options for this rate limiter will be a combination of what's in [TokenBucketRateLimiterOptions](https://github.com/dotnet/runtime/blob/315549e9ece8ff5d5b04316a62c1b73f7370cff2/src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/TokenBucketRateLimiterOptions.cs) and [RedisCacheOptions](https://github.com/dotnet/aspnetcore/blob/29f8aa30564aebf3365b57011498212831715b64/src/Caching/StackExchangeRedis/src/RedisCacheOptions.cs).

We might want to make it easier to reuse existing `IConnectionMultiplexer` since there will likely be multiple instances by default. Another way to avoid unnecessary Redis connections might be to implement `PartitionedRateLimiter<string>` instead of `RateLimiter`.

@BrennanConroy @wtgodbe

### Additional context

Customers are [asking for this.](https://github.com/dotnet/aspnetcore/issues/37380#issuecomment-1137813130).

**Comment by adityamandaleeka** [comment]
Worth considering this scenario when working on this too: https://github.com/dotnet/aspnetcore/issues/41861

**Comment by Expecho** [comment]
Is this now moved to .Net 9 or what is the current status?

**Comment by amcasey** [comment]
@mgravell Does this align with the Redis work you're planning?

**Comment by ReubenBond** [comment]
BTW, I have proof-of-concept implementations here, which might serve as inspiration (or example of what not to do - either way is fine to me): https://github.com/ReubenBond/DistributedRateLimiting.Redis/tree/main/DistributedRateLimiting.Redis

**Comment by mgravell** [comment]
@amcasey this is certainly a topic I've discussed with Aditya, but this isn't what I'm focusing on right now; I have thoughts on it, in particular around avoiding latency overheads via broadcast, but that also depends on how strict we need to be ("N per T" is very different depending on whether there's an implied "give-or-take" hand-wave in there), but I haven't actively spiked anything. I'll take a look at Reuben's braindump.

**Comment by md-redwan-hossain** [comment]
Is there any progress of this feature?

**Comment by adityamandaleeka** [comment]
Hey @md-redwan-hossain, we haven't worked more on this since the last update. We're almost certainly not going to have time for this in the .NET 9 timeframe.

We're definitely aware of the need for a better distributed story for rate limiting.

**Comment by stylianosnicoletti** [comment]
Any chance of adding an SQL Database (Oracle) as distributed cache for rate limiting?

**Comment by mgravell** [comment]
If/when we add something here, there would absolutely be some open abstraction that arbitrary backend implementations could provide, in the same way as we do today for `IDistributedCache` which has a wide range of implementations. If you're asking about SQL, then: is (MSFT) providing SQL Server: not necessarily hard. PostgreSQL and sqlite I would by default *assume* would be community provided, but it isn't impossible for us to look at them. Based on my experience fighting server backends for Dapper, Oracle tends to be trickier due to server setup and licensing (plus, to be honest, familiarity of the teams). I *suspect* Oracle would be lower on the list of things MSFT would provide an implementation for, but again: there's no reason why the community (yourself, even) couldn't provide that implementation in a NuGet package - especially if you have the relevant servers, skills, and practical need (which, being realistic, tends to result in better implementations, due to more realistic testing etc)

**Comment by clement911** [comment]
I would like to +1 this feature.
Our azure app service loads balances incoming requests between all instances. We need to apply a system wide rate limit rather than a per-machine rate limit. I'm sure it's a very common scenario.

Stripe has published a detailed blog post on how they implemented the token bucket algo in reddis lua:
https://gist.github.com/ptarjan/e38f45f2dfe601419ca3af937fff574d

**Comment by sscowden** [comment]
I agree this would be a really useful feature in .NET and Azure. I'd use this tomorrow if available
